### PR TITLE
Bug 1381623 - A new Photon Menu using the AS context menu.

### PR DESCRIPTION
**Pull request by farhanpatel**
Here is the initial Photon Menu 
pasted image at 2017_08_01 02_53 pm

**There are some problems this menu is trying to solve.**
- Depending on the state of the browser the menu has different items. This creates an unpredictable experience for the user. 
- There are multiple ways of doing things. Creating larger menus than are needed. (Settings appears in tab tray and in BVC)
- The grid makes it hard to skim the options. You cant just scan the list quickly. Your eyes have to jump around. And this is especially hard when the menu items keep changing based on if you are in a home panel or website.
- The grid menu is gone from Desktop. 
 
**This menu does create some problems**
- The menu is too large on small devices. 
- The animation isnt the greatest. 
- Its different 

There will be a lot of iteration on the design. We need to improve the animation and the reachability of the menu. The goal is to land a functional menu and work from there.

Technically, I decided to reuse the PhotonActionMenu instead of repurposing the MenuViewController because there is a lot of logic in MVC tied to laying out in a grid that is no longer needed. It was easier to reuse the AS context menu and build it with 2 states rather than re purpose MVC. 
I've used a Protocol approach to creating the menu actions which keeps most of the code in `PhotonActionSheetProtocol` which removes a few hundred lines from BVC.
I've also stripped out AppMenuState in many places, Because we no longer change the menu based on the App's state we dont need it anymore.

The HomePage button can no longer be in the toolbar menu. Because the share action has been repurposed. 

This patch depends on the other UI changes and should land _after_ those changes have landed. 
There will be followup work and I'll create the bugs for that. They are tracked [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1381619)

Here are the things that will be done after
- use localized strings for menu options
- Tab tray button. (kinda shoved in there but people _really_ want this)
- Possibly moving the QR code button out of the menu and into the URL bar
- updating the animation
- tweaking the order of menu items. 
- updating icons.

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
TabTrayController.swift | 436 | Limit vertical whitespace to a single empty line. Currently 2. 
TabTrayController.swift | 425 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
PhotonActionSheetAnimator.swift | 80 | Limit vertical whitespace to a single empty line. Currently 2. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Client/Frontend/Browser/TabTrayController.swift [hunk 1]
```diff
@@ -240,13 +236,10 @@ class TabTrayController: UIViewController {
         let toolbar = TrayToolbar()
         toolbar.addTabButton.addTarget(self, action: #selector(TabTrayController.SELdidClickAddTab), for: .touchUpInside)
         toolbar.maskButton.addTarget(self, action: #selector(TabTrayController.SELdidTogglePrivateMode), for: .touchUpInside)
+        toolbar.menuButton.addTarget(self, action: #selector(TabTrayController.didTapMenu), for: .touchUpInside)
```
Comment (jhugman) [inline Client/Frontend/Browser/TabTrayController.swift]: We need to either get rid of the `SEL` prefixes of the other selectors, or use it for this method.

File: Client/Application/QuickActions.swift [hunk 2]
```diff
@@ -115,7 +115,7 @@ class QuickActions: NSObject {
                 handleOpenURL(withBrowserViewController: browserViewController, urlToOpen: urlToOpen)
             }
         case .qrCode:
-            handleQRCode(withBrowserViewController: browserViewController)
+            handleQRCode(with: browserViewController)
```
Comment (jhugman) [inline Client/Application/QuickActions.swift]: Nice.

File: Client/Application/QuickActions.swift [hunk 3]
```diff
@@ -133,7 +133,10 @@ class QuickActions: NSObject {
         bvc.switchToTabForURLOrOpen(urlToOpen, isPrivileged: true)
     }
     
-    fileprivate func handleQRCode(withBrowserViewController bvc: BrowserViewController) {
-        bvc.openQRReader()
+    fileprivate func handleQRCode(with vc: QRCodeViewControllerDelegate & UIViewController) {
```
Comment (jhugman) [inline Client/Application/QuickActions.swift]: I'm ambivalent about using algebraic types here. 

But standing ovation for using it to get rid of BVC.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 4]
```diff
@@ -711,6 +694,15 @@ extension TabTrayController: SettingsDelegate {
     }
 }
 
+extension TabTrayController: PhotonActionSheetProtocol {
+    func didTapMenu() {
+        let closeAllTabs = PhotonActionSheetItem(title: "Close All Tabs", iconString: "action_remove") { _ in
```
Comment (jhugman) [inline Client/Frontend/Browser/TabTrayController.swift]: Nit: use already localized `Strings.swift` string.

(There will already be a localized string, which if not in `Strings.swift` should be moved there).

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            popoverVC.delegate = viewController
+            popoverVC.sourceView = view
+            popoverVC.sourceRect = CGRect(x: view.frame.width/2, y: view.frame.size.height * 0.75, width: 1, height: 1)
+            popoverVC.permittedArrowDirections = UIPopoverArrowDirection.up
+        }
+        viewController.present(sheet, animated: true, completion: nil)
+    }
+    
+    //Returns a list of actions which is used to build a menu
+    //parameter OpenURL is a clousre that can open a given URL in some view controller. It is up to the class using the menu to know how to open the url
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Nit: spelling.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    func getHomePanelActions(openURL: @escaping (URL) -> Void, vcDelegate: PageOptionsVC) -> [PhotonActionSheetItem] {
+        let openQR = PhotonActionSheetItem(title: "QR Scanner", iconString: "menu-ScanQRCode") { action in
+            let qrCodeViewController = QRCodeViewController()
+            qrCodeViewController.qrCodeDelegate = vcDelegate
+            let controller = UINavigationController(rootViewController: qrCodeViewController)
+            vcDelegate.present(controller, animated: true, completion: nil)
+        }
+        
+        let openSettings = PhotonActionSheetItem(title: "Settings", iconString: "menu-Settings") { action in
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Nit: use existing localized string.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            popoverVC.sourceRect = CGRect(x: view.frame.width/2, y: view.frame.size.height * 0.75, width: 1, height: 1)
+            popoverVC.permittedArrowDirections = UIPopoverArrowDirection.up
+        }
+        viewController.present(sheet, animated: true, completion: nil)
+    }
+    
+    //Returns a list of actions which is used to build a menu
+    //parameter OpenURL is a clousre that can open a given URL in some view controller. It is up to the class using the menu to know how to open the url
+    func getHomePanelActions(openURL: @escaping (URL) -> Void, vcDelegate: PageOptionsVC) -> [PhotonActionSheetItem] {
+        let openQR = PhotonActionSheetItem(title: "QR Scanner", iconString: "menu-ScanQRCode") { action in
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Nit: use existing localized string.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            settingsTableViewController.tabManager = self.tabManager
+            settingsTableViewController.settingsDelegate = vcDelegate
+            
+            let controller = SettingsNavigationController(rootViewController: settingsTableViewController)
+            controller.popoverDelegate = vcDelegate
+            controller.modalPresentationStyle = UIModalPresentationStyle.formSheet
+            vcDelegate.present(controller, animated: true, completion: nil)
+        }
+        
+        let openTopSites = PhotonActionSheetItem(title: Strings.AppMenuTopSitesTitleString, iconString: "menu-panel-TopSites") { action in
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: I see what is happening here, and I'm not sure I like it. 

Putting the actual action with the menu item seems appealing to start with, but quickly becomes difficult when you want to perform the same action without going through the menu. e.g. `findInPage`, `bookmark` can be triggered from multiple places, and you want the identical code to be triggered for each. 

This was part of the rationale of trying to get a redux pattern in place.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+     These items repersent global options that are presented in the menu
+     TODO: These icons should all have the icons and use Strings.swift
+     */
+    
+    typealias PageOptionsVC = QRCodeViewControllerDelegate & SettingsDelegate & PresentingModalViewControllerDelegate & UIViewController
+    
+    func getOtherPanelActions() -> [PhotonActionSheetItem] {
+        
+        let noImageEnabled = NoImageModeHelper.isActivated(profile.prefs)
+        let noImageText = "Hide Images: \(noImageEnabled ? "On" : "Off")"
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Nit: we already have localized strings for this. 

Aside for less experienced readers: you should never concatenate localized strings.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        let noImageMode = PhotonActionSheetItem(title: noImageText, iconString: "menu-NoImageMode", isEnabled: noImageEnabled) { action in
+            NoImageModeHelper.toggle(profile: self.profile, tabManager: self.tabManager)
+        }
+        
+        let nightModeEnabled = NightModeHelper.isActivated(profile.prefs)
+        let nightModeText = "Night Mode: \(nightModeEnabled ? "On" : "Off")"
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Nit: we already have localized strings for this.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+                       findInPage:  @escaping () -> Void,
+                       presentableVC: PresentableVC) -> Array<[PhotonActionSheetItem]> {
+        
+        let toggleActionTitle = tab.desktopSite ? Strings.AppMenuViewMobileSiteTitleString : Strings.AppMenuViewDesktopSiteTitleString
+        let toggleDesktopSite = PhotonActionSheetItem(title: toggleActionTitle, iconString: "menu-RequestDesktopSite") { action in
+            tab.toggleDesktopSite()
+        }
+        
+        let setHomePage = PhotonActionSheetItem(title: Strings.AppMenuSetHomePageTitleString, iconString: "menu-Home") { action in
+            //TODO: pass a VC. this doesnt _need_ to be a HomePageHelper
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Unsure I understand this comment.

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+                userData[QuickActions.TabTitleKey] = title
+            }
+            QuickActions.sharedInstance.addDynamicApplicationShortcutItemOfType(.openLastBookmark,
+                                                                                withUserData: userData,
+                                                                                toApplication: UIApplication.shared)
+            tab.isBookmarked = true
+        }
+        
+        let removeBookmark = PhotonActionSheetItem(title: Strings.AppMenuRemoveBookmarkTitleString, iconString: "menu-Bookmark-Remove") { action in
+            //TODO: can all this logic go somewhere else?
```
Comment (jhugman) [inline Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: Yes, the action needs to be disassociated from where it's triggered.

**Pull request merged**